### PR TITLE
[confmap] Mark `confmap.strictlyTypedInput` as stable

### LINKED ISSUE #10552: Tracking issue for `confmap.strictlyTypedInput` feature gate
This issue describes the `confmap.strictlyTypedInput` feature gate, intended to address #8565 and #9532.

### What is changing?

1. Configurations relying on the implicit type casting behaviors listed on #9532 will start to fail.
2. Configurations using URI expansion (i.e. `field: ${env:ENV}`) for string-typed fields will use the value passed in `ENV` verbatim without intermediate type casting.

### What does it mean to me?

Starting on v0.105.0, you may have to adapt your configuration to use the correct type on some fields. This should be a straightforward. You are able to skip the error and update your configuration at your own pace by disabling the `confmap.strictlyTypedInput` feature gate.

Please **reply below** if:
- You are unsure what an error means or how to update your configuration
- An error tells you that you can disable the feature gate to skip it, but this does not work. 

### Changelog 

| Collector version | Change | PR |
|-------------------|--------|----|
|    v0.103.0               |  Introduce feature gate as **alpha**      |  #10400  |
|    v0.105.0              |   Promote feature gate to **beta** (enabled by default)     |  #10554  |
|    *v0.108.0*              |   Promote feature gate to **stable** (can't be disabled)     |  #10793  |
|    *v0.110.0*              |   Remove feature gate (errors out when referenced)     |    |

### PR DESCRIPTION
**Pull request by mx-psi**
#### Description

Marks `confmap.strictlyTypedInput` as stable.

#### Link to tracking issue

Fixes #10552

Blocked by:
- #10794
- #10795

### CODE REVIEW

File: .chloggen/mx-psi_strict-typing-stable.yaml
Comment (songy23) [inline .chloggen/mx-psi_strict-typing-stable.yaml:4]: Shouldn't this be breaking?

Comment (mx-psi) [reply .chloggen/mx-psi_strict-typing-stable.yaml]: Yes, fixed, thanks!

**Review by songy23 — APPROVED** [review]
LGTM

**Review by TylerHelmuth — APPROVED** [review]

**Comment by mx-psi** [comment]
Discussed on the stability meeting from 2024-08-07, we can have this for v0.108.0 instead of v0.107.0

**Comment by TylerHelmuth** [comment]
@mx-psi ready for a rebase

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
I've run into an interesting egde case. Investigating now. Please don't merge for now

File: otelcol/configprovider.go
Comment (codeboten) [inline otelcol/configprovider.go:111]: I'm likely missing context here, but in other places in this PR we've kept the behaviour inside the check if the `StrictlyTypedInputGate.IsEnabled` and here it's removed.

**Review by codeboten — COMMENTED** [review]
Thanks @mx-psi, just one question

File: otelcol/configprovider.go
Comment (mx-psi) [reply otelcol/configprovider.go:111]: The coda says:

```
Hint: Temporarily restore the previous behavior by disabling 
      the `confmap.strictlyTypedInput` feature gate. More details at:
      https://github.com/open-telemetry/opentelemetry-collector/issues/10552
```

which is not actionable after this PR is merged (you cannot disable the feature gate). Should we keep some sort of custom message? It could make sense to give people a place where they can complain

**Comment by mx-psi** [comment]
> I've run into an interesting edge case. Investigating now. Please don't merge for now

:+1:, if you find something, could you file a separate issue?

**Comment by mx-psi** [comment]
I found one issue that should be fixed here #10897

**Comment by mx-psi** [comment]
Last call before I merge this cc @open-telemetry/collector-approvers.

I checked the two real issues reported on #10552 and verified that these were fixed with the binary:

<details>
<summary>Can't pass a number to a string field</summary>

**Issue**: https://github.com/open-telemetry/opentelemetry-collector/issues/10552#issuecomment-2237383731 (can't pass a number to a `string` field)

**Config used to reproduce the issue**:
```
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: ${env:ENDPOINT}

exporters:
  nop:

service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [nop]
```

**Logs with 0.107.0**:
```
❯ ENDPOINT=1234 ./otelcol --config config.yaml 
Error: failed to get config: cannot unmarshal the configuration: decoding failed due to the following error(s):

error decoding 'receivers': error reading configuration for "otlp": decoding failed due to the following error(s):

error decoding '': decoding failed due to the following error(s):

'protocols.grpc.endpoint' expected type 'string', got unconvertible type 'int', value: '1234'

Hint: Temporarily restore the previous behavior by disabling 
      the `confmap.strictlyTypedInput` feature gate. More details at:
      https://github.com/open-telemetry/opentelemetry-collector/issues/10552
2024/08/19 18:31:22 collector server run finished with error: failed to get config: cannot unmarshal the configuration: decoding failed due to the following error(s):

error decoding 'receivers': error reading configuration for "otlp": decoding failed due to the following error(s):

error decoding '': decoding failed due to the following error(s):

'protocols.grpc.endpoint' expected type 'string', got unconvertible type 'int', value: '1234'

Hint: Temporarily restore the previous behavior by disabling 
      the `confmap.strictlyTypedInput` feature gate. More details at:
      https://github.com/open-telemetry/opentelemetry-collector/issues/10552
```

**Logs with HEAD** (fails to start but passes validation and unmarshaling):
```
❯ ENDPOINT=1234 ./bin/otelcorecol_linux_amd64 --config config.yaml 
2024-08-19T18:18:35.255+0200    info    service@v0.107.0/service.go:114 Setting up own telemetry...
2024-08-19T18:18:35.255+0200    info    service@v0.107.0/telemetry.go:98        Serving metrics {"address": ":8888", "metrics level": "Normal"}
2024-08-19T18:18:35.258+0200    info    service@v0.107.0/service.go:193 Starting otelcorecol... {"Version": "0.107.0-dev", "NumCPU": 20}
2024-08-19T18:18:35.258+0200    info    extensions/extensions.go:37     Starting extensions...
2024-08-19T18:18:35.258+0200    info    otlpreceiver@v0.107.0/otlp.go:103       Starting GRPC server    {"kind": "receiver", "name": "otlp", "data_type": "traces", "endpoint": "1234"}
2024-08-19T18:18:35.258+0200    error   graph/graph.go:430      Failed to start component       {"error": "listen tcp: address 1234: missing port in address", "type": "Receiver", "id": "otlp"}
2024-08-19T18:18:35.258+0200    info    service@v0.107.0/service.go:256 Starting shutdown...
2024-08-19T18:18:35.258+0200    info    extensions/extensions.go:64     Stopping extensions...
2024-08-19T18:18:35.258+0200    info    service@v0.107.0/service.go:270 Shutdown complete.
Error: cannot start pipelines: listen tcp: address 1234: missing port in address
2024/08/19 18:18:35 collector server run finished with error: cannot start pipelines: listen tcp: address 1234: missing port in address
```

</details>

<details>
<summary>TLS min version in env variable</summary>

**Issue**: https://github.com/open-telemetry/opentelemetry-collector/issues/10552#issuecomment-2284953792

**Config used to reproduce the issue**:
```
receivers:
  nop:

exporters:
  otlphttp:
    endpoint: localhost:4317
    tls:
      min_version: ${env:TLS_MIN_VERSION}

service:
  pipelines:
    traces:
      receivers: [nop]
      exporters: [otlphttp]
```

**Logs with 0.107.0**:
```
❯ TLS_MIN_VERSION=1.2 ./otelcol --config config.yaml 
Error: failed to get config: cannot unmarshal the configuration: decoding failed due to the following error(s):

error decoding 'exporters': error reading configuration for "otlphttp": decoding failed due to the following error(s):

'tls.min_version' expected type 'string', got unconvertible type 'float64', value: '1.2'

Hint: Temporarily restore the previous behavior by disabling 
      the `confmap.strictlyTypedInput` feature gate. More details at:
      https://github.com/open-telemetry/opentelemetry-collector/issues/10552
2024/08/19 18:32:18 collector server run finished with error: failed to get config: cannot unmarshal the configuration: decoding failed due to the following error(s):

error decoding 'exporters': error reading configuration for "otlphttp": decoding failed due to the following error(s):

'tls.min_version' expected type 'string', got unconvertible type 'float64', value: '1.2'

Hint: Temporarily restore the previous behavior by disabling 
      the `confmap.strictlyTypedInput` feature gate. More details at:
      https://github.com/open-telemetry/opentelemetry-collector/issues/10552
```

**Logs with HEAD**:
```
❯ TLS_MIN_VERSION=1.2 ./bin/otelcorecol_linux_amd64 --config config.yaml 
2024-08-19T18:26:25.705+0200    info    service@v0.107.0/service.go:114 Setting up own telemetry...
2024-08-19T18:26:25.706+0200    info    service@v0.107.0/telemetry.go:98        Serving metrics {"address": ":8888", "metrics level": "Normal"}
2024-08-19T18:26:25.709+0200    info    service@v0.107.0/service.go:193 Starting otelcorecol... {"Version": "0.107.0-dev", "NumCPU": 20}
2024-08-19T18:26:25.709+0200    info    extensions/extensions.go:37     Starting extensions...
2024-08-19T18:26:25.710+0200    info    service@v0.107.0/service.go:219 Everything is ready. Begin running and processing data.
2024-08-19T18:26:25.710+0200    info    localhostgate/featuregate.go:63 The default endpoints for all servers in components have changed to use localhost instead of 0.0.0.0. Disable the feature gate to temporarily revert to the previous default.     {"feature gate ID": "component.UseLocalHostAsDefaultHost"}
^C2024-08-19T18:26:27.819+0200  info    otelcol@v0.107.0/collector.go:318       Received signal from OS {"signal": "interrupt"}
2024-08-19T18:26:27.819+0200    info    service@v0.107.0/service.go:256 Starting shutdown...
2024-08-19T18:26:27.819+0200    info    extensions/extensions.go:64     Stopping extensions...
2024-08-19T18:26:27.819+0200    info    service@v0.107.0/service.go:270 Shutdown complete.
```

</details>

I have looked into adding end to end tests, I think we can and should do that, but it will take some time and I don't want to block this and confmap 1.0 on these tests.

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
